src/main/initVerticalMod.F90

```diff
@@ -257,24 +255,50 @@ subroutine initVertical(bounds, glc_behavior, snow_depth, thick_wall, thick_roof
 
        zisoi(0) = 0._r8
        do j = 1, nlevgrnd-1
-       zisoi(j) = 0.5_r8*(zsoi(j)+zsoi(j+1))         !interface depths
+          zisoi(j) = 0.5_r8*(zsoi(j)+zsoi(j+1))         !interface depths
        enddo
        zisoi(nlevgrnd) = zsoi(nlevgrnd) + 0.5_r8*dzsoi(nlevgrnd)
 
-    else if ( soil_layerstruct == '49SL_10m' ) then
-       !scs: 10 meter soil column, nlevsoi set to 49 in clm_varpar
-       do j = 1,10
-          dzsoi(j)= 1.e-2_r8     !10mm layers
-       enddo
-       do j = 11,19
-          dzsoi(j)= 1.e-1_r8     !100 mm layers
-       enddo
-       do j = 20,nlevsoi+1       !300 mm layers
-          dzsoi(j)= 3.e-1_r8
-       enddo
-       do j = nlevsoi+2,nlevgrnd !10 meter bedrock layers
-          dzsoi(j)= 10._r8
-       enddo
+    else if (calc_method == 'thickness-based') then
+       if (soil_layerstruct_predefined == '49SL_10m') then
+          !scs: 10 meter soil column, nlevsoi set to 49 in clm_varpar
+          do j = 1, 10
+             dzsoi(j) = 1.e-2_r8     ! 10-mm layers
+          enddo
+          do j = 11, 19
+             dzsoi(j) = 1.e-1_r8     ! 100-mm layers
+          enddo
+          do j = 20, nlevsoi+1       ! 300-mm layers
+             dzsoi(j) = 3.e-1_r8
+          enddo
+          do j = nlevsoi+2,nlevgrnd  ! 10-m bedrock layers
+             dzsoi(j) = 10._r8
+          enddo
+       else if (soil_layerstruct_predefined == '20SL_8.5m') then
+          do j = 1, 4  ! linear increase in layer thickness of...
+             dzsoi(j) = j * 0.02_r8                     ! ...2 cm each layer
+          enddo
+          do j = 5, 13
+             dzsoi(j) = dzsoi(4) + (j - 4) * 0.04_r8    ! ...4 cm each layer
+          enddo
+          do j = 14, nlevsoi
+             dzsoi(j) = dzsoi(13) + (j - 13) * 0.10_r8  ! ...10 cm each layer
+          enddo
+          do j = nlevsoi + 1, nlevgrnd  ! bedrock layers
+             dzsoi(j) = dzsoi(nlevsoi) + (((j - nlevsoi) * 25._r8)**1.5_r8) / 100._r8
+          enddo
+       else if (soil_layerstruct_predefined == '4SL_2m') then
+          dzsoi(1) = 0.1_r8
+          dzsoi(2) = 0.3_r8
+          dzsoi(3) = 0.6_r8
+          dzsoi(4) = 1.0_r8
+          dzsoi(5) = 1.0_r8
+       else if (soil_layerstruct_userdefined(1) /= rundef) then
+          do j = 1, nlevgrnd
+             ! read dzsoi from user-entered namelist vector
+             dzsoi(j) = soil_layerstruct_userdefined(j)
+          end do
+       end if  ! thickness-based options
```

Comment:
It would be good to have a couple of extra conditions in this big if/else if block, basically mirroring some of the checking you do in clm_varctl:
- Before checking userdefined, you should have a block that checks `else if (soil_layerstruct_predefined /= 'UNSET')`, and abort with a message "Unrecognized soil layer structure"
- Finally, you should have an `else` which would be entered if soil_layerstruct_predefined is UNSET and soil_layerstruct_userdefined(1) is undefined: abort in this case, too.

I know you already have checking along these lines in clm_varctl, but the extra checking here would help catch the possibility that the two get out of sync.

(A more ideal situation would be if the code in this module could be combined with the code in clm_varctl, but I'm suspecting that may be difficult because of when each of them needs to be called.)

Reply:
The if-block here and the if-block in clm_varpar are a bit different. The former distinguishes by calc_method (node-based vs. thickness-based). The latter distinguishes by whether user-defined and/or predefined have valid entries. However, the former is preceded by the if-block that sets calc_method. This one already mostly handles everything:
- assigns `node-based` for two predefined cases
- assigns `thickness-based` for three predefined and all user-defined cases
- aborts if predefined and user-defined cases were not valid
For completeness, I am adding error checks to this if-block in case both predefined and user-defined are set.

In the if-block that you brought up, I'm adding an error check in case we slip past the if-block that sets calc_method without a valid calc_method.